src/main/java/com/google/cloud/tools/app/impl/cloudsdk/internal/process/DefaultProcessRunner.java

```diff
@@ -25,38 +26,41 @@
  * monitoring output and checking the exit code of the child process.
  */
 public class DefaultProcessRunner implements ProcessRunner {
-
-  private ProcessBuilder processBuilder;
-  private boolean async = false;
-  private ProcessOutputLineListener stdOutLineListener;
-  private ProcessOutputLineListener stdErrLineListener;
+  private boolean async;
+  private List<ProcessOutputLineListener> stdOutLineListeners;
+  private List<ProcessOutputLineListener> stdErrLineListeners;
   private ProcessExitListener exitListener;
 
+  private ProcessBuilder processBuilder = new ProcessBuilder();
   private Process process;
 
-  /**
-   * Create the process runner with a default process builder, with inheritIO enabled.
-   */
-  public DefaultProcessRunner() {
-    this(new ProcessBuilder());
-    processBuilder.inheritIO();
-  }
+  private Map<String, String> environment;
 
-  /**
-   * Create the process runner with the provided process builder.
-   */
-  public DefaultProcessRunner(ProcessBuilder processBuilder) {
-    this.processBuilder = processBuilder;
+  private DefaultProcessRunner(Builder builder) {
+    this.async = builder.async;
+    this.stdOutLineListeners = builder.stdOutLineListeners;
+    this.stdErrLineListeners = builder.stdErrLineListeners;
+    this.exitListener = builder.exitListener;
   }
 
   /**
    * Executes a shell command.
    *
+   * <p>If any output listeners were configured, output will go to them only. Otherwise, process
+   * output will be redirected to the caller via inheritIO.
+   *
    * @param command The shell command to execute
    */
   public void run(String[] command) throws ProcessRunnerException {
     try {
-
+      // configure process builder
+      final ProcessBuilder processBuilder = new ProcessBuilder();
+      if (stdErrLineListeners.size() == 0 && stdOutLineListeners.size() == 0) {
```

Comment:
I feel like inheriting should be per stream?
processBuilder.redirectError(Redirect.INHERIT) ?